packages/migrations/src/batched-migrations/batched-migration-runner.ts

```diff
@@ -0,0 +1,200 @@
+import {
+  loadSqlEquiv,
+  queryAsync,
+  queryValidatedOneRow,
+  queryValidatedSingleColumnOneRow,
+  queryValidatedZeroOrOneRow,
+} from '@prairielearn/postgres';
+import { logger } from '@prairielearn/logger';
+import { serializeError } from 'serialize-error';
+import { z } from 'zod';
+
+import {
+  BatchedMigrationStatus,
+  BatchedMigrationRow,
+  updateBatchedMigrationStatus,
+  BatchedMigrationStatusSchema,
+  BatchedMigrationImplementation,
+} from './batched-migration';
+import {
+  BatchedMigrationJobRowSchema,
+  BatchedMigrationJobStatus,
+  BatchedMigrationJobRow,
+} from './batched-migration-job';
+
+const sql = loadSqlEquiv(__filename);
+
+interface BatchedMigrationRunnerOptions {
+  logProgress?: boolean;
+}
+
+export class BatchedMigrationRunner {
+  private options: BatchedMigrationRunnerOptions;
+  private migration: BatchedMigrationRow;
+  private migrationImplementation: BatchedMigrationImplementation;
+  private migrationStatus: BatchedMigrationStatus;
+
+  constructor(
+    migration: BatchedMigrationRow,
+    migrationImplementation: BatchedMigrationImplementation,
+    options: BatchedMigrationRunnerOptions = {}
+  ) {
+    this.options = options;
+    this.migration = migration;
+    this.migrationImplementation = migrationImplementation;
+    this.migrationStatus = migration.status;
+  }
+
+  private log(message: string, ...meta: any[]) {
+    if (this.options.logProgress) {
+      logger.info(`[${this.migration.filename}] ${message}`, ...meta);
+    }
+  }
+
+  private async hasIncompleteJobs(migration: BatchedMigrationRow): Promise<boolean> {
+    return queryValidatedSingleColumnOneRow(
+      sql.batched_migration_has_incomplete_jobs,
+      { batched_migration_id: migration.id },
+      z.boolean()
+    );
+  }
+
+  private async hasFailedJobs(migration: BatchedMigrationRow): Promise<boolean> {
+    return queryValidatedSingleColumnOneRow(
+      sql.batched_migration_has_failed_jobs,
+      { batched_migration_id: migration.id },
+      z.boolean()
+    );
+  }
+
+  private async refreshMigrationStatus(migration: BatchedMigrationRow) {
+    this.migrationStatus = await queryValidatedSingleColumnOneRow(
+      sql.get_migration_status,
+      {
+        id: migration.id,
+      },
+      BatchedMigrationStatusSchema
+    );
+  }
+
+  private async finishRunningMigration(migration: BatchedMigrationRow) {
+    // Safety check: if there are any pending jobs, don't mark this
+    // migration as finished.
+    if (await this.hasIncompleteJobs(migration)) {
+      this.log(`Incomplete jobs found, not marking as finished`);
+    }
+
+    const hasFailedJobs = await this.hasFailedJobs(migration);
+    const finalStatus = hasFailedJobs ? 'failed' : 'succeeded';
+    await updateBatchedMigrationStatus(migration.id, finalStatus);
+    this.log(`Finished with status '${finalStatus}'`);
```

Comment:
Should the code below be in an 'else' block?

Reply:
Yep! I accidentally dropped a `return` during a refactor. Restored in `c709898`.